ansible_collections/arista/avd/tests/unit/test/test_defined_plugin.py

```diff
@@ -0,0 +1,117 @@
+                self.defined_function(VALUE, fail_action=FAIL_ACTION,
+                                      var_name=VAR_NAME, test_value=TEST_VALUE, warn_msg=warn_msg)
+            elif str(FAIL_ACTION).lower() == 'error':
+                if VAR_NAME is not None:
+                    err_msg = f"{VAR_NAME} was set to {VALUE} but we expected {TEST_VALUE}!"
+                else:
+                    err_msg = f"A variable was set to {VALUE} but we expected {TEST_VALUE}!"
+                self.defined_function(VALUE, fail_action=FAIL_ACTION,
+                                      var_name=VAR_NAME, test_value=TEST_VALUE, err_msg=err_msg)
+
+    @pytest.mark.parametrize("VALUE", VALUE_LIST)
+    @pytest.mark.parametrize("FAIL_ACTION", FAIL_ACTION_LIST)
+    @pytest.mark.parametrize("VAR_NAME", VAR_NAME_LIST)
+    @pytest.mark.parametrize("VAR_TYPE", VAR_TYPE_LIST)
+    def test_defined_plugin_var_type(self, VALUE, FAIL_ACTION, VAR_NAME, VAR_TYPE):
+        type_list = ['float', 'int', 'str', 'list', 'dict', 'tuple', 'bool']
+        if not isinstance(VALUE, Undefined) and VALUE is not None:
+            if str(VAR_TYPE).lower() in type_list and str(VAR_TYPE).lower() != type(VALUE).__name__:
+                if str(FAIL_ACTION).lower() == 'warning':
+                    if VAR_NAME is not None:
+                        warn_msg = f"{VAR_NAME} was a {type(VALUE).__name__} but we expected a {str(VAR_TYPE).lower()}. Output may be incorrect or incomplete!"
+                    else:
+                        warn_msg = f"A variable was a {type(VALUE).__name__} but we expected a {str(VAR_TYPE).lower()}. Output may be incorrect or incomplete!"
+                    self.defined_function(
+                        VALUE, fail_action=FAIL_ACTION, var_name=VAR_NAME, var_type=VAR_TYPE, warn_msg=warn_msg)
+                elif str(FAIL_ACTION).lower() == 'error':
+                    if VAR_NAME is not None:
+                        err_msg = f"{VAR_NAME} was a {type(VALUE).__name__} but we expected a {str(VAR_TYPE).lower()}!"
+                    else:
+                        err_msg = f"A variable was a {type(VALUE).__name__} but we expected a {str(VAR_TYPE).lower()}!"
+                    self.defined_function(
+                        VALUE, fail_action=FAIL_ACTION, var_name=VAR_NAME, var_type=VAR_TYPE, err_msg=err_msg)
+
+    @pytest.mark.parametrize("VALUE", VALUE_LIST)
+    @pytest.mark.parametrize("TEST_VALUE", TEST_VALUE_LIST)
+    @pytest.mark.parametrize("INVALID_FAIL_ACTION", INVALID_FAIL_ACTION_LIST)
+    def test_defined_plugin_fail_action_None(self, VALUE, TEST_VALUE, INVALID_FAIL_ACTION):
+        resp = defined(VALUE, test_value=TEST_VALUE, fail_action=INVALID_FAIL_ACTION)
+        if not isinstance(VALUE, Undefined) and VALUE is not None:
+            if TEST_VALUE is not None and VALUE != TEST_VALUE:
+                assert resp is False
+            else:
+                assert resp is True
+        else:
+            assert resp is False
```

Comment:
Please test the below scenario:

```
    @pytest.mark.parametrize("VALUE", VALUE_LIST)
    @pytest.mark.parametrize("VAR_TYPE", VAR_TYPE_LIST)
    def test_defined_plugin_var_type_fail_action_None(self, VALUE, VAR_TYPE):
        type_list = ['float', 'int', 'str', 'list', 'dict', 'tuple', 'bool']
        if not isinstance(VALUE, Undefined) and VALUE is not None:
            resp = defined(VALUE, var_type=VAR_TYPE)
            if str(VAR_TYPE).lower() in type_list and type(VALUE).__name__ != str(VAR_TYPE).lower():
                assert resp is False
            else:
                assert resp is True
```

Reply:
Tested this test case with different var_type and values.